File: straight_skeleton/src/main/java/util/TreeCheck.java
package at.tugraz.igi.util;

import java.util.*;
import org.jgrapht.*;
import org.jgrapht.graph.*;

public class TreeCheck {


    /**
     * Checks if the graph defined by given set of points and lines is a tree.
     *
     * Points can be null.
     *
     * TODO: how far do we need to relax this for polygons with holes?
     */
    public static boolean isTree(List<Point> points, List<Line> edges) throws Exception {
        Pseudograph<Point, DefaultEdge> gr = new Pseudograph<>(DefaultEdge.class);
        if (points != null)
          for (Point p: points)
              gr.addVertex(p);
        for (Line e: edges) {
            if (!gr.containsVertex(e.getP1()))
              gr.addVertex(e.getP1());
            if (!gr.containsVertex(e.getP2()))
              gr.addVertex(e.getP2());
            gr.addEdge(e.getP1(), e.getP2());
        }
        for (Point p: points) {
            if(gr.degreeOf(p) != 1)
                throw new Exception("Bad skeleton: a vertex is not a leaf in skeleton");
        }
        return GraphTests.isTree(gr);
    }

}

File: kotton/src/main/kotlin/geometry/treecheck.kt
package  madhat.kotton.geometry

import madhat.kotton.utils.*
import javax.vecmath.*
import org.jgrapht.*
import org.jgrapht.graph.*

class TreeCheckException(override var message:String): Exception(message)

/**
 * Checks if given set of edges is a tree (we only care for simple holeless polygons)
 * and if every vertex of the input is a leaf in it (we check only in this one direction).
 *
 * Throws exception describing the error if not.
 */
fun runTreeCheck(input: ParsedPolygon,
                 edges: Set<Pair<Point2d, Point2d>>) {

    var graph: Pseudograph<Point2d, DefaultEdge> = Pseudograph(DefaultEdge::class.java)
    for (e in edges) {
        val p = e.first
        val q = e.second
        for (pt in listOf(p, q)) {
            if (!graph.containsVertex(pt)) graph.addVertex(pt)
        }
        graph.addEdge(p, q)
    }

    for ((i, v) in input.coordinates)
        if (graph.degreeOf(v) != 1)
            throw TreeCheckException("Bad skeleton: a vertex ($i) is not a leaf (has degree ${graph.degreeOf(v)})")

    if(!GraphTests.isTree(graph))
        throw TreeCheckException("Bad skeleton: not a tree")

}
